sonar-xml-plugin/src/main/java/org/sonar/plugins/xml/highlighting/XMLHighlighting.java

```diff
@@ -0,0 +1,191 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.sonar.plugins.xml.highlighting;
+
+import javax.xml.stream.Location;
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLStreamConstants;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamReader;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+public class XMLHighlighting {
+
+  private XMLHighlighting() {
+  }
+
+  public static List<HighlightingData> getHighlightingData(File file, Charset charset) throws IOException, XMLStreamException {
+    return getHighlightingData(new FileInputStream(file), new SourceFileOffsets(file, charset));
+  }
+
+  public static List<HighlightingData> getHighlightingData(String content) throws XMLStreamException {
+    InputStream inputStream = new ByteArrayInputStream(content.getBytes(StandardCharsets.UTF_8));
+    return getHighlightingData(inputStream, new SourceFileOffsets(content));
+  }
+
+  public static List<HighlightingData> getHighlightingData(InputStream inputStream, SourceFileOffsets sourceFileOffsets) throws XMLStreamException {
+    XMLStreamReader streamReader = XMLInputFactory.newInstance().createXMLStreamReader(inputStream);
+
+    List<HighlightingData> highlighting = new ArrayList<>();
+
+    if (sourceFileOffsets.content().startsWith("<?xml")) {
+      highlighting.add(new HighlightingData(0, 5, "k"));
+      int closingBracketStartOffset = getClosingBracketStartOffset(0, sourceFileOffsets.content());
+      highlighting.add(new HighlightingData(closingBracketStartOffset - 1, closingBracketStartOffset + 1, "k"));
+
+      highlightAttributes(highlighting, 5, closingBracketStartOffset, sourceFileOffsets.content());
+    }
+
+    while (streamReader.hasNext()) {
+      Location prevLocation = streamReader.getLocation();
+      streamReader.next();
+      int startOffset = sourceFileOffsets.startOffset(streamReader.getLocation().getLineNumber(), streamReader.getLocation().getColumnNumber());
+
+      switch (streamReader.getEventType()) {
+        case XMLStreamConstants.START_ELEMENT:
+          int closingBracketStartOffset = getClosingBracketStartOffset(startOffset, sourceFileOffsets.content());
+          int endOffset = startOffset + getNameWithNamespaceLength(streamReader) + 1;
+          highlightAttributes(highlighting, endOffset, closingBracketStartOffset, sourceFileOffsets.content());
+
+          highlighting.add(new HighlightingData(startOffset, endOffset, "k"));
+          highlighting.add(new HighlightingData(closingBracketStartOffset, closingBracketStartOffset + 1, "k"));
+          break;
+
+        case XMLStreamConstants.END_ELEMENT:
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this switch case number of lines from 10 to at most 5, for example by extracting code into methods. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1151)